package.json

```
@@ -45,7 +45,7 @@
     "postcss-discard-comments": "^7.0.3",
     "postcss-nesting": "^13.0.1",
     "prettier": "^3.3.3",
-    "puppeteer": "23.3.1",
+    "puppeteer": "^23.9.0",
```

```suggestion
    "puppeteer": "23.9.0",
```

I know that @Snuffleupagus asked whether the range can be more specific at https://github.com/mozilla/pdf.js/pull/19115#discussion_r1860797639 with the following question:

> Can we "relax" the version requirements now, since we only locked it down previsouly because of the integration-test failures?

The answer is No, we should not relax the version. Puppeteer hardcodes browser versions and bumps their versions once in a while. Every browser update may potentially result in rendering differences, and consequently result in reftest failures.

If we were to use imprecise versions here, then a PDF.js PR may pass tests today and fail tomorrow, without any changes to the code in the PR. That is a debugging nightmare.

Therefore, we should continue to lock to specific Puppeteer versions and bump the version once in a while, and update all reftests at once.

---

> If we were to use imprecise versions here, then a PDF.js PR may pass tests today and fail tomorrow, without any changes to the code in the PR. That is a debugging nightmare.

I don't believe that's an issue though, since we have a `package-lock.json` file that *should* still guarantee that the versions cannot just change randomly!?

---

I think so too. `package.json` only defines a range, and `package-lock.json` pins it to a fixed version for a reproducible environment. The `^` merely indicates that we accept the mentioned version and higher minor versions because those _should_, according to semantic versioning, keep working. We previously only pinned it in `package.json` because we found out that newer versions would actually break, so in that case there was a specific use case, but in general we want to encourage updating to newer versions.

---

The use of package-lock.json would indeed avoid the scenario I mentioned before (PRs breaking), if `npm ci` is used to install dependencies, which is indeed the case, e.g
 as seen at https://github.com/mozilla/botio-files-pdfjs/blob/5f24d5c1fee89fdf91cf5a8e73d5f1058439ad75/on_cmd_browsertest.js#L6

But unless someone bumps the versions in `package-lock.json`, CI will NOT use recent browser versions. This seems not a concern based on your comments, particularly:

> Note that bumping versions is already done explicitly using PRs,

Assuming that you are referring to bumping versions in package-lock.json. If you are not regularly bumping puppeteer versions in package-lock.json, then tests will be run with older browser versions, without test coverage in the latest browser versions.

---

Yes, this analysis is correct. We are regularly (usually bi-weekly) bumping versions in `package-lock.json`. Puppeteer was the only exception for this process because of #18773, but this PR should fix that issue so that it'll be taken along in the regular version bump rounds.